narwhal/worker/src/batch_maker.rs

```diff
@@ -274,6 +274,18 @@ impl BatchMaker {
         let metadata = batch.versioned_metadata().clone();
 
         Some(Box::pin(async move {
+            let responses = responses;
+
+            // Also wait quorum broadcast here.
+            //
+            // Error can only happen when the worker is shutting down.
+            // All other errors, e.g. timeouts, failure responses from individual peers,
+            // are retried indefinitely underneath.
+            if broadcasted_to_quorum.await.is_err() {
+                // Drop all response handlers to signal error.
+                return;
+            }
+
             // Now save it to disk
             let digest = batch.digest();
```

Comment:
Shouldn't we save the batch to disk first before doing anything else? If the broadcast does fail due to worker shutting down, we still want the node to come back up and process the batch that others may have received already.

Reply:
I don't think batches are re-read from the batch store on restart. Batches are only read when other workers request them, when verifying headers or fetching for certificates. Storing the batch after quorum broadcast is slightly better because a batch does not finish quorum broadcast will not be in a header anyway, but this should be minor.

Reply:
Aah you are right it is not part of a header yet